### DISCUSSION TITLE
Impossible to stop a too long job or startup event

### DISCUSSION THREAD

**Discussion by Xmaster6y** [fastapi/fastapi#10135] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import logging
import time

from fastapi import FastAPI

app = FastAPI()

@app.on_event("startup")
def startup_event():
    logger = logging.getLogger("uvicorn")
    logger.info("Starting up...")
    while True:
        logger.info("long process running...")
        time.sleep(1)

@app.get("/")
async def home():
    while True:
        logger.info("other long process running...")
        time.sleep(1)
    return {"success": "Welcome to the API!"}
```

### Description

I happen to have a long startup event waiting for other microservices to set up.  I noticed that it was not possible to stop the process with a usual ^C or using signals. It is really not handy when it comes to testing (manual and auto).

I looked into the lifespan part of the doc but couldn't find something similar.

To reproduce, simply run the code with `uvicorn main:app --port 8000`. if you can stop it, I might have other problems. Even when removing the startup infinite loop, I cannot stop the request process.

I may be missing something big, but this feels weird.

### Operating System

Linux

### FastAPI Version

0.101.1

### Pydantic Version

2.2.1

### Python Version

3.9.16

**Comment by JavierSanchezCastro** [comment]
Looking at the logs I see that this msg appears after @app.on_event("startup"), so I suppose CTRL+C is only available after "startup".

INFO:     Application startup complete
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)

A possible workaround that I just tested is using "--reload" flag, which should work.

**Reply by Xmaster6y** [reply]
Thanks for the info, I am aware of the logs, but I don't really understand why it is blocked during startup.

The `--reload` doesn't solve my problem, seeing that this option is just there to watch file modifications.

**Reply by JavierSanchezCastro** [reply]
I think it's uvicorn side, check this https://github.com/encode/uvicorn/issues/1301

About the ``--reload``, I see that in Ubuntu it's not working, but does in Windows, I'm not really sure if it's uvicorn or how the system handles the signals.

**Comment by YuriiMotov** [comment, marked as answer]
As JavierSanchezCastro mentioned, it comes from Uvicorn (https://github.com/encode/uvicorn/issues/1301).

The solution from [here](https://github.com/encode/uvicorn/issues/1301#issuecomment-1020971978) is to set the handler for `SIGINT` signal:

```python
import asyncio
from contextlib import asynccontextmanager
import logging
import signal
import sys

from fastapi import FastAPI

logger = logging.getLogger("uvicorn")

def receive_signal(signalNumber, frame):
    print("Received:", signalNumber)
    sys.exit()

@asynccontextmanager
async def lifespan(app: FastAPI):
    signal.signal(signal.SIGINT, receive_signal)
    logger.info("Starting up...")
    for i in range(10):
        logger.info(f"long process running ({i})...")
        await asyncio.sleep(1)
    yield
    logger.info("Shutdown...")


app = FastAPI(lifespan=lifespan)
```
